utils/near-rate-limiter/src/framed_read.rs

```diff
@@ -0,0 +1,354 @@
+        self.total_sizeof_messages_in_progress.fetch_add(msg_size, Ordering::SeqCst);
+    }
+
+    // Decrease limits by size of the message and notify ThrottledFramedReader to try to
+    // read again
+    pub fn remove_msg(&mut self, msg_size: usize) {
+        self.num_messages_in_progress.fetch_sub(1, Ordering::SeqCst);
+        self.total_sizeof_messages_in_progress.fetch_sub(msg_size, Ordering::SeqCst);
+
+        // Notify throttled framed reader
+        self.semaphore.add_permits(1);
+    }
+}
+
+impl<T, D> ThrottledFrameRead<T, D>
+where
+    T: AsyncRead,
+    D: Decoder,
+{
+    /// Creates a new `NearFramedRead` with the given `decoder`.
+    pub fn new(
+        inner: T,
+        decoder: D,
+        throttle_controller: ThrottleController,
+        semaphore: PollSemaphore,
+    ) -> ThrottledFrameRead<T, D> {
+        ThrottledFrameRead {
+            inner: FramedImpl {
+                inner,
+                codec: decoder,
+                state: Default::default(),
+                throttle_controller,
+                semaphore,
+            },
+        }
+    }
+}
+
+impl<T, D> ThrottledFrameRead<T, D> {
+    pub fn get_ref(&self) -> &T {
+        &self.inner.inner
+    }
+
+    pub fn into_inner(self) -> T {
+        self.inner.inner
+    }
+
+    pub fn decoder(&self) -> &D {
+        &self.inner.codec
+    }
+
+    pub fn read_buffer(&self) -> &BytesMut {
+        &self.inner.state.buffer
+    }
+}
+
+// This impl just defers to the underlying FramedImpl
+impl<T, D> Stream for ThrottledFrameRead<T, D>
+where
+    T: AsyncRead,
+    D: Decoder,
+{
+    type Item = Result<D::Item, D::Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.project().inner.poll_next(cx)
+    }
+}
+
+pub(crate) struct ReadFrame {
+    pub(crate) eof: bool,
+    pub(crate) is_readable: bool,
+    pub(crate) buffer: BytesMut,
+}
+
+impl Default for ReadFrame {
+    fn default() -> Self {
+        Self { eof: false, is_readable: false, buffer: BytesMut::with_capacity(INITIAL_CAPACITY) }
+    }
+}
+
+impl<T, U, R> Stream for FramedImpl<T, U, R>
+where
+    T: AsyncRead,
+    U: Decoder,
+    R: BorrowMut<ReadFrame>,
+{
+    type Item = Result<U::Item, U::Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut pinned = self.project();
+        let state: &mut ReadFrame = pinned.state.borrow_mut();
+        loop {
+            if !pinned.throttle_controller.is_ready() {
+                // This will cause us to subscribe to notifier when something gets pushed to
+                // `pinned.receiver`. If there is an element in the queue, we will check again.
+                ready!(PollSemaphore::poll_next(Pin::new(pinned.semaphore), cx));
+            }
+            if !pinned.throttle_controller.is_ready() {
+                return Poll::Pending;
+            }
```

Comment:
Hm, I don't understand how this branch works. It is checked on the same event loop iteration as the one before, with identical condition, so, logically, it's just redundant? If we do end up hitting this `return Pendng`, than we don't have anything scheduled to wake us up, no?